### PR TITLE
Added RenameHandler, Tests, Integration tests, certificate verification

### PR DESCRIPTION
**Pull request by antooa**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
+FROM golang:1.11.4-alpine3.8 as builder
+
+EXPOSE 8080
+
+ENV GOPATH /go
+
+RUN apk add make
+
+RUN apk add openssl ca-certificates
```
[the remark this answers is not in the document]
Comment (antooa) [reply Dockerfile]: Left it in the last stage only.

File: cmd/server/main.go [hunk 2]
```diff
@@ -2,10 +2,13 @@ package main
...
 import (
 	srv "GoExcercise/server"
+	"fmt"
 	"log"
+	"os"
 )
 
 func main() {
 	server := srv.NewFileServer()
+	_, _ = fmt.Fprintf(os.Stderr, "Listen On"+server.Addr)
```
[the remark this answers is not in the document]
Comment (antooa) [reply cmd/server/main.go]: Done.

File: handler/handler.go [hunk 3]
```diff
@@ -3,40 +3,75 @@ package handler
...
+		writer.WriteHeader(500)
+		return
+	}
+	if _, err := os.Stat(DownloadFolder + newName); !os.IsNotExist(err) {
+		writer.WriteHeader(409)
+		return
+	}
+	err := os.Rename(DownloadFolder + oldName, DownloadFolder + newName)
+	if err != nil{
+		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done.

File: handler/handler.go [hunk 4]
```diff
@@ -71,12 +106,20 @@ func UploadHandler(writer http.ResponseWriter, request *http.Request) {
 
 	uri := request.FormValue("uri")
 	fileName := namegen.GenerateFileName(10)
-	err := UploadFile(path.Join(downloadFolder, fileName), uri)
+	err := UploadFile(path.Join(DownloadFolder, fileName), uri)
 	if err != nil {
-		_, _ = io.WriteString(writer, err.Error())
+		_, err = writer.Write([]byte(err.Error()))
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done.

File: handler/handler.go [hunk 5]
```diff
@@ -3,40 +3,75 @@ package handler
...
-	if _, err := os.Stat(downloadFolder + filename); os.IsNotExist(err) {
+	if _, err := os.Stat(DownloadFolder + filename); os.IsNotExist(err) {
 		writer.WriteHeader(404)
 		return
 	}
-	err := os.Remove(downloadFolder + filename)
+
+	err := os.Remove(DownloadFolder + filename)
 	if err != nil{
 		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done.

File: handler/handler.go [hunk 6]
```diff
@@ -46,11 +81,11 @@ func DeleteHandler(writer http.ResponseWriter, request *http.Request) {
...
 
-	if _, err := os.Stat(downloadFolder + filename); os.IsNotExist(err) {
+	if _, err := os.Stat(DownloadFolder + filename); os.IsNotExist(err) {
 		writer.WriteHeader(404)
 		return
 	}
-	file, err := os.Open(downloadFolder + filename)
+	file, err := os.Open(DownloadFolder + filename)
 	if err != nil{
 		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done.

File: handler/handler_test.go [hunk 7]
```diff
@@ -1,32 +1,122 @@
...
-}
\ No newline at end of file
+	//err = os.Remove("testfile1")
+	//if err != nil {
+	//	log.Fatal(err)
+	//}
+	//err = os.Remove("testfile2")
+	//if err != nil {
+	//	log.Fatal(err)
+	//}
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler_test.go]: Done.

File: handler/handler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rvolykh) [reply handler/handler.go]: it is not done ^

File: server/file-server_test.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+func TestUploadDelete(t *testing.T) {
+	u := url.URL{
+		Scheme:   "http",
+		Host:     "localhost:8080",
+		Path:     "/upload",
+		RawQuery: url.Values{"uri": []string{"https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_120x44dp.png"}}.Encode(),
+	}
+	resp, err := http.Get(u.String())
+	if err != nil {
+		t.Errorf("Couldn't GET : %v; Got %v", u.String(), err)
```
[the remark this answers is not in the document]
Comment (antooa) [reply server/file-server_test.go]: I should replace all t.Error with t.Fatal? Did I got it right?

[the remark this answers is not in the document]
Comment (rvolykh) [reply server/file-server_test.go]: t.Error => t.Fatal
t.Errorf => t.Fatalf

File: server/file-server_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	delUrl := url.URL{
+		Scheme:   "http",
+		Host:     "localhost:8080",
+		Path:     "/delete/"+newName,
+	}
+
+	deleteResp, err := http.Get(delUrl.String())
+	if err != nil {
+		t.Errorf("Couldn't GET : %v", delUrl.String())
+	}
```
[the remark this answers is not in the document]
Comment (antooa) [reply server/file-server_test.go]: Done

File: server/file-server_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	renUrl := url.URL{
+		Scheme:   "http",
+		Host:     "localhost:8080",
+		Path:     "/rename/"+filename+"/new/"+newName,
+	}
+	renResp, err := http.Get(renUrl.String())
+	if err != nil {
+		t.Errorf("Couldn't GET : %v", renUrl.String())
+	}
+
```
[the remark this answers is not in the document]
Comment (antooa) [reply server/file-server_test.go]: Done

File: server/file-server_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (antooa) [reply server/file-server_test.go]: Done

File: handler/handler.go [hunk 11]
```diff
@@ -46,20 +84,20 @@ func DeleteHandler(writer http.ResponseWriter, request *http.Request) {
...
-	if _, err := os.Stat(downloadFolder + filename); os.IsNotExist(err) {
+	if _, err := os.Stat(DownloadFolder + filename); os.IsNotExist(err) {
 		writer.WriteHeader(404)
 		return
 	}
-	file, err := os.Open(downloadFolder + filename)
+	file, err := os.Open(DownloadFolder + filename)
 	if err != nil{
-		log.Fatal(err)
+		http.Error(writer, err.Error(), 500)
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,23 @@
+FROM golang:1.11.4-alpine3.8 as builder
+
+EXPOSE 8080
```
[the remark this answers is not in the document]
Comment (antooa) [reply Dockerfile]: Moved

File: handler/handler.go [hunk 13]
```diff
@@ -3,41 +3,79 @@ package handler
...
+	if _, err := os.Stat(DownloadFolder + filename); os.IsNotExist(err) {
 		writer.WriteHeader(404)
 		return
 	}
-	err := os.Remove(downloadFolder + filename)
+
+	err := os.Remove(DownloadFolder + filename)
 	if err != nil{
-		log.Fatal(err)
+		http.Error(writer, err.Error(), 500)
```
[the remark this answers is not in the document]
Comment (antooa) [reply handler/handler.go]: Done

**Pull request merged**